File: jp/scripts/compare_i18n_keys.rb
require 'yaml'

def flatten(data, prefix='')
  if data.is_a?(Hash)
    data.flat_map do |key, value|
      new_prefix = prefix.empty? ? key.to_s : "#{prefix}.#{key}"
      flatten(value, new_prefix)
    end
  else
    [[prefix, data]]
  end
end

de = YAML.load_file('_i18n/de.yml')
en = YAML.load_file('_i18n/en.yml')
de_keys = flatten(de).to_h
en_keys = flatten(en).to_h
missing = de_keys.keys - en_keys.keys
extra = en_keys.keys - de_keys.keys
puts "MISSING #{missing.size}"
missing.sort.each do |k|
  puts k
  puts '---'
  puts de_keys[k]
  puts
end
puts "EXTRA #{extra.size}"
extra.sort.each do |k|
  puts k
  puts '---'
  puts en_keys[k]
  puts
end

File: en/scripts/compare_i18n_keys.py
import yaml
from pathlib import Path

def flatten(data, prefix=''):
    if isinstance(data, dict):
        for key, value in data.items():
            new_prefix = f"{prefix}.{key}" if prefix else key
            yield from flatten(value, new_prefix)
    else:
        yield prefix, data

files = [('_i18n/de.yml', '_i18n/en.yml')]
for de_path, en_path in files:
    de = yaml.safe_load(Path(de_path).read_text(encoding='utf-8'))
    en = yaml.safe_load(Path(en_path).read_text(encoding='utf-8'))
    de_keys = {k:v for k,v in flatten(de)}
    en_keys = {k:v for k,v in flatten(en)}
    missing = sorted(k for k in de_keys if k not in en_keys)
    extra = sorted(k for k in en_keys if k not in de_keys)
    print('MISSING', len(missing))
    for k in missing:
        print(k)
        print('---')
        print(de_keys[k])
        print()
    print('EXTRA', len(extra))
    for k in extra:
        print(k)
        print('---')
        print(en_keys[k])
        print()
